capsules/extra/src/net/ipv6/ipv6_recv.rs

```diff
@@ -71,7 +71,8 @@ impl<'a> SixlowpanRxClient for IP6RecvStruct<'a> {
         }
         match IP6Header::decode(buf).done() {
             Some((offset, ip6_header)) => {
-                let checksum_result = ip6_header.check_transport_checksum(&buf[offset..len]);
+                let checksum_result =
+                    ip6_header.check_transport_checksum(&buf[offset..offset + len]);
```

Comment:
I'm a little confused by this, why is it valid to index into buf beyond the received `len`?

Reply:
The `len` field is only the length of the ip6_header. `offset` is where the header begins so `offset+len` constitutes the entire packet. 

For non-encrypted 15.4 packets, the index for len extends beyond the passed buf. This has to do with the way the offsets are calculated in the framer. This comment from a previous implementation addressed the issue of the index being too long, but reasoned that it was okay. I originally looked into why this was occurring, but could not find a satisfactory answer. I copied the comment below. I'll look some more into this. 

Previous Comment from `framer.rs`:
    ```
                // Note: there is a complication here regarding the offsets.
                // When the received frame has security enabled, the payload
                // (including the payload IEs) is encrypted, and hence the data
                // payload field includes the encrypted payload IEs too.
                // However, when the frame is not encrypted, the data payload
                // field does not include the payload IEs.
                //
                // This is fine because we re-parse the unsecured frame before
                // exposing it to the user. At that time, the data payload field
                // will not include the payload IEs.
```

Reply:
I am certainly no expert on this stack. I tried to find where `SixlowpanRxClient` is used, and I found it in sixlowpan_state.rs. 

The receive callback provides dgram_size:

https://github.com/tock/tock/blob/d068b0549e847430e79637c4d94ca125cf06c343/capsules/extra/src/net/sixlowpan/sixlowpan_state.rs#L775-L779

Which seems to primarily be set by start_receive:

https://github.com/tock/tock/blob/d068b0549e847430e79637c4d94ca125cf06c343/capsules/extra/src/net/sixlowpan/sixlowpan_state.rs#L924-L930

It seems like if you are allowed to read beyond that len then the api is bad and we should fix that, or change the calling side.